fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Utils.java

```diff
@@ -536,4 +569,273 @@ private static <T extends ScalarOperator> void collect(ScalarOperator root, Clas
+        StatementBase mvStmt;
+        try {
+            List<StatementBase> statementBases =
+                    com.starrocks.sql.parser.SqlParser.parse(sql, connectContext.getSessionVariable());
+            Preconditions.checkState(statementBases.size() == 1);
+            mvStmt = statementBases.get(0);
+        } catch (ParsingException parsingException) {
+            LOG.warn("parse sql:{} failed", sql, parsingException);
+            return null;
+        }
+        Preconditions.checkState(mvStmt instanceof QueryStatement);
+        Analyzer.analyze(mvStmt, connectContext);
+        QueryRelation query = ((QueryStatement) mvStmt).getQueryRelation();
+        LogicalPlan logicalPlan =
+                new RelationTransformer(columnRefFactory, connectContext).transformWithSelectLimit(query);
+        // optimize the sql by rule and disable rule based materialized view rewrite
+        OptimizerConfig optimizerConfig = new OptimizerConfig(OptimizerConfig.OptimizerAlgorithm.RULE_BASED);
+        // TODO: disable mv query rewrite rules
+        Optimizer optimizer = new Optimizer(optimizerConfig);
+        OptExpression optimizedPlan = optimizer.optimize(
+                connectContext,
+                logicalPlan.getRoot(),
+                new PhysicalPropertySet(),
+                new ColumnRefSet(logicalPlan.getOutputColumn()),
+                columnRefFactory);
+        return Pair.create(optimizedPlan, logicalPlan);
+    }
+
+    public static void updatePartialPartitionPredicate(MaterializedView mv, ColumnRefFactory columnRefFactory,
+                                                 Set<String> partitionsToRefresh, OptExpression mvPlan) {
+        // to support partial partition rewrite
+        PartitionInfo partitionInfo = mv.getPartitionInfo();
+        if (!(partitionInfo instanceof ExpressionRangePartitionInfo)) {
+            return;
+        }
+        ExpressionRangePartitionInfo exprPartitionInfo = (ExpressionRangePartitionInfo) partitionInfo;
+        if (partitionsToRefresh.isEmpty()) {
+            // all partitions are uptodate, do not add filter
+            return;
+        }
+
+        Set<Long> outdatePartitionIds = Sets.newHashSet();
+        for (Partition partition : mv.getPartitions()) {
+            if (partitionsToRefresh.contains(partition.getName())) {
+                outdatePartitionIds.add(partition.getId());
+            }
+        }
+        if (outdatePartitionIds.size() == mv.getPartitions().size()) {
+            // all partitions are out of date
+            // should not reach here, it will be filtered when registering mv
+            return;
+        }
+        // now only one column is supported
+        Column partitionColumn = exprPartitionInfo.getPartitionColumns().get(0);
+        List<Range<PartitionKey>> uptodatePartitionRanges = exprPartitionInfo.getRangeList(outdatePartitionIds, false);
```

Comment:
outdatePartitionIds get getRangeList could be uptodatePartitionRanges?

Reply:
no, outdatePartitionIds is excluded, getRangeList will return all partitions that are not in outdatePartitionIds